### PR TITLE
Make most set operations not in-place

### PR DESCRIPTION
**Pull request by elazarg**
All binary set operations are now non-mutating. This is makes the interfaces more consistent, simplifies reasoning in some cases, and sometimes saves lines of code.

I'm missing a lint directive that warns about unused return values.

### CODE REVIEW

File: pkg/interval/intervalset.go [hunk 1]
```diff
@@ -38,9 +38,7 @@ func (c *CanonicalSet) Equal(other *CanonicalSet) bool {
 	return true
 }
 
-// AddInterval adds a new interval range to the set
-func (c *CanonicalSet) AddInterval(v Interval) {
-	set := c.IntervalSet
+func addInterval(v Interval, set []Interval) []Interval {
```
Comment (zivnevo) [inline pkg/interval/intervalset.go:41]: What is the motivation behind this change?
Why have 2 `addInterval` functions when you can have one?

File: pkg/ipblock/ipblock.go [hunk 2]
```diff
@@ -184,14 +185,13 @@ func addIntervalToList(ipbNew *IPBlock, ipbList []*IPBlock) []*IPBlock {
 		if !ipb.ipRange.Overlaps(ipbNew.ipRange) {
```
Comment (zivnevo) [inline pkg/ipblock/ipblock.go:185]: Something to add to function renaming: 'Overlaps' should change to 'Overlap'

File: pkg/ipblock/ipblock.go [hunk 3]
```diff
@@ -236,19 +242,15 @@ func FromCidrList(cidrsList []string) (*IPBlock, error) {
...
 }
 
-// FromCidrExcept returns an IPBlock object from input cidr str an exceptions cidr str
-func FromCidrExcept(cidr string, exceptions []string) (*IPBlock, error) {
-	span, err := cidrToInterval(cidr)
-	if err != nil {
-		return nil, err
-	}
-	ipRange := interval.FromInterval(span.Start, span.End)
+// Except returns an IPBlock object from input cidr str an exceptions cidr str
```
Comment (zivnevo) [inline pkg/ipblock/ipblock.go:245]: Fix doc

File: pkg/ipblock/ipblock.go [hunk 4]
```diff
@@ -236,19 +242,15 @@ func FromCidrList(cidrsList []string) (*IPBlock, error) {
...
+func (b *IPBlock) Except(exceptions ...string) (*IPBlock, error) {
+	ipRange := b.ipRange.Copy()
 	for i := range exceptions {
 		intervalHole, err := cidrToInterval(exceptions[i])
 		if err != nil {
 			return nil, err
 		}
-		ipRange.AddHole(intervalHole)
+		ipRange = ipRange.Subtract(interval.FromInterval(intervalHole.Start, intervalHole.End))
```
Comment (zivnevo) [inline pkg/ipblock/ipblock.go:253]: Code might be clearer if you create IPBlocks from the cidrs and call `IPBlock.Subtract`

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ipblock/ipblock.go (hunk 4 above)
Comment (elazarg) [reply pkg/ipblock/ipblock.go:253]: Also more efficient, since the copy only happens once. Done.

File: pkg/ipblock/ipblock.go (hunk 3 above)
Comment (elazarg) [reply pkg/ipblock/ipblock.go]: Changed to
> Except creates a new IP block that excludes the specified CIDRs from the current IP block

File: pkg/interval/intervalset.go (hunk 1 above)
Comment (elazarg) [reply pkg/interval/intervalset.go]: Just an experiment. Reverted.

File: pkg/ipblock/ipblock.go [hunk 5]
```diff
@@ -78,8 +78,9 @@ func (b *IPBlock) Intersect(c *IPBlock) *IPBlock {
 	if b == c {
 		return res
 	}
```
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/ipblock/ipblock.go]: Done

**Review by zivnevo — APPROVED** [review]
LGTM
